### PR TITLE
add new geo_distance UDF to compute the haversine distance between 2 …

### LINKED ISSUE #353: allowing geolocation queries
is it planned to integrate and allow geolocation queries  when you have columns with lat lon ... in the  stream  with some  new  keyword in the ksql dml langage

### PR DESCRIPTION
**Pull request by blueedgenick**
…lat-lon points. Fixes #353

### Description 
Added a new KUDF to compute the haversine distance between 2 points, plus corresponding syntax guide update. (original request in #353)

### Testing done 
New unit tests for positive, negative, and boundary cases for the new KUDF.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/syntax-reference.rst [hunk 1]
```diff
@@ -838,6 +838,12 @@ Scalar functions
 +------------------------+------------------------------------------------------------+---------------------------------------------------+
 | FLOOR                  |  ``FLOOR(col1)``                                           | The floor of a value                              |
 +------------------------+------------------------------------------------------------+---------------------------------------------------+
+| GEO_DISTANCE           | ``GEODISTANCE(lat1, lon1,                                  | Distance between two lat-lon points, |
```
Comment (miguno) [inline docs/syntax-reference.rst]: The table syntax is broken because the ending `|` must be aligned to the rest of the table.

Think:

```
| foo         |
| geo         |
```

And not:

```
| foo         |
| geo    |
```

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/geo/GeoDistance.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+/**
+ * Compute the distance between two points on the surface of the earth, according to the Haversine
+ * formula for "great circle distance". The 2 input points should be specified as (lat, lon) pairs,
+ * measured in decimal degrees.
+ * 
+ * <p>An optional fifth parameter allows to specify either "MI" (miles) or "KM" (kilometers) as the
+ * desired unit for the output measurement. Default is KM.
+ *
+ */
+public class GeoDistance implements Kudf {
```
Comment (miguno) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/geo/GeoDistance.java]: I think that, by our convention, `GeoDistance` must be named `GeoDistanceKudf`.

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/geo/GeoDistanceTest.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
+package io.confluent.ksql.function.udf.geo;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import io.confluent.ksql.function.KsqlFunctionException;
+import static org.junit.Assert.assertEquals;
+
+public class GeoDistanceTest {
```
Comment (miguno) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/geo/GeoDistanceTest.java]: We should also validate that the KUDF fails when less params than 4 or more params than 5 are given to it.

**Comment by miguno** [comment]
Thanks a lot for creating this PR @blueedgenick !

File: docs/syntax-reference.rst (hunk 1 above)
Comment (blueedgenick) [reply docs/syntax-reference.rst]: ahh - got it. i needed a much wider monitor to actually see this :)  - fixed.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/geo/GeoDistance.java (hunk 2 above)
Comment (blueedgenick) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udf/geo/GeoDistance.java]: changed as suggested

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/geo/GeoDistanceTest.java (hunk 3 above)
Comment (blueedgenick) [reply ksql-engine/src/test/java/io/confluent/ksql/function/udf/geo/GeoDistanceTest.java]: new test added for this. seems like a candidate for a more generic test which does this same thing for all registered UDFs ? the FunctionRegistry already has all the info that would be required to do this...

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/geo/GeoDistanceKudf.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    }
+  }
+
+  private double selectEarthRadiusToUse(Object... args) {
+    double chosenRadius = EARTH_RADIUS_KM;
+    if (args.length == 5) {
+      String outputUnit = args[4].toString().toLowerCase();
+      if (outputUnit.equals("mi") || outputUnit.startsWith("mile")) {
+        chosenRadius = EARTH_RADIUS_MILES;
+      } else if (outputUnit.equals("km") || outputUnit.startsWith("kilomet")) {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/geo/GeoDistanceKudf.java]: nit: `equals("kilometeres")` ?
similarly above for `mile`. Only as otherwise the valid values for this could `kilometasdfhaklsjfhak` which makes no sense!

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/geo/GeoDistanceKudfTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        (double) distanceUdf.evaluate(37.4439, -122.1663, 51.5257, -0.1122, "KM"), 0.5);
+    assertEquals(5365.66, (double) distanceUdf.evaluate(37.4439, -122.1663, 51.5257, -0.1122, "MI"),
+        0.5);
+  }
+
+  @Test
+  public void shouldFailWithTooFewParams() {
+    expectedException.expect(KsqlFunctionException.class);
+    expectedException.expectMessage("GeoDistance function expects either 4 or 5 arguments");
+    assertEquals(8634.6528, (double) distanceUdf.evaluate(37.4439, -122.1663), 0.5);
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/geo/GeoDistanceKudfTest.java]: here and elsewhere, `assertEquals` doesn't make sense as you are expecting an exception

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/geo/GeoDistanceKudf.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    if ((args.length < 4) || (args.length > 5)) {
+      throw new KsqlFunctionException(
+          "GeoDistance function expects either 4 or 5 arguments: lat1, lon1, lat2, lon2, (MI/KM)");
+    }
+
+    double lat1 = ((Number) args[0]).doubleValue();
+    double lon1 = ((Number) args[1]).doubleValue();
+    double lat2 = ((Number) args[2]).doubleValue();
+    double lon2 = ((Number) args[3]).doubleValue();
+    checkParameterValuesAreLegal(lat1, lon1, lat2, lon2);
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/geo/GeoDistanceKudf.java]: nit: we usually call these things `validateX`. So `validateLatLong` would be nicer.

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/geo/GeoDistanceKudfTest.java (hunk 5 above)
Comment (blueedgenick) [reply ksql-engine/src/test/java/io/confluent/ksql/function/udf/geo/GeoDistanceKudfTest.java]: good point! will remove...

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/geo/GeoDistanceKudf.java (hunk 6 above)
Comment (blueedgenick) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udf/geo/GeoDistanceKudf.java]: sure, can change accordingly.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/geo/GeoDistanceKudf.java (hunk 4 above)
Comment (blueedgenick) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udf/geo/GeoDistanceKudf.java]: so my thinking here was 2-fold: first up, i wanted to avoid having to force people to remember whether it's singular or plural distance units (similar to how we do for time units, for example); secondarily, to accommodate folks in places where they like to write "kilometre" and not force them to use the anglicized spelling all the time. Would you guys prefer this be changed to test for all the allowable permutations explicitly ?

**Comment by apurvam** [comment]
Looks like the build is failing with 
```
UDQ4WZVR2JNESS2I2K5FOJFVB66LKQCKXVKQVV2BUMCKYVG2J3LA@tmp/withMaven79fbf958/maven-spy-20180514-174813-2698602095392567274176.log
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.1:compile (default-compile) on project ksql-engine: Compilation failure
[ERROR] /home/****/workspace/onfluentinc-pr_ksql_PR-1227-UDQ4WZVR2JNESS2I2K5FOJFVB66LKQCKXVKQVV2BUMCKYVG2J3LA/ksql-engine/src/main/java/io/confluent/ksql/physical/AddTimestampColumn.java:[43,7] method does not override or implement a method from a supertype
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :ksql-engine
[Pipeline] }
```
You may need to rebase onto master and try again.

**Comment by apurvam** [comment]
@blueedgenick it would be good to fix up the git history and rebase onto master so that this can be merged.

**Comment by blueedgenick** [comment]
Closing in favor of cherry-picking the original commits into new PR #1356
